### ISSUE TITLE
useEffect that works with DOM/UI

### ISSUE THREAD

**Issue by WebReflection** [getify/TNG-Hooks#21]
I understand the general purpose of this library (pretty identical to [augmentor](https://github.com/WebReflection/augmentor#augmentor)) but `useEffect` as it is makes no sense for anything UI related, since it requires machinery around.

[dom-augmentor](https://github.com/WebReflection/dom-augmentor#dom-augmentor) uses _augmentor_ flexibility to provide such usability, is TNG also planning to offer DOM/UI related hooks that work when nodes are connected or disconnected * from the DOM?

`*` connected and disconnected as in Custom Elements definition

**Comment by getify** [comment]
The future direction of TNG (as described in #20) should be entirely compatible with using effects for UI related tasks. I think it positions TNG as a very low-level piece of the puzzle, free to let all the DOM/UI orchestration to be layered on top of it.

Quoting an example from that thread:

```js
function hitCounter(btnID) {
   // accessing the current TNG hooks-context
   var hooksContext = this;

   var [count,updateCount] = useState(0);
   var elem = useRef();

   useEffect(function onInit(){
      elem.current = document.getElementById(btnID);
      elem.current.addEventListener("click",function onClick(){
         updateCount(x => x + 1);

         // re-render the button
         hitCounter(hooksContext,btnID).effects();
      },false);
   },[btnID]);

   useEffect(function onRender(){
      elem.value = count;
   },[count]);
}

hitCounter = TNG(hitCounter);

hitCounter("the-btn").effects();   // button initially says: "0"

// click the button, it now says: "1"
// cilck the button again, it now says: "2"
// cilck the button yet again, it now says: "3"
```

So, let's break down a few things about what's happening here.

First, notice the call to `.effects()` near the bottom of the snippet. If that call isn't made right away, none of the effects will run. Effects will be lazy, meaning that outside of TNG there's more control over deciding when effects should run. I ran them right away for simplicity of illustration, but that's certainly
not required.

In this example, the component re-renders itself -- see the nested `hitCounter(..)` call in the middle of the snippet -- each time the button is clicked. But that's not required, just an illustration. You could have DOM/UI orchestration similar to React that decides, based on state changes and effects, when a component should re-render itself.

Right now, there's just one bucket of effects (`useEffect(..)`), and they're all lazy, meaning they all are deferred until you call `.effects()` on the returned context object, at which point they all run.

BUT... this whole change is happening so that there will be the possibility to introduce other sorts of effects, such as those which should run only on the next animation frame, or whatever. So we'll eventually introduce other effects hooks like `useWHATEVER(..)`, and then orchestration outside of TNG could decide when each different bucket of events needs to run, according to whatever UI/DOM approach is desired.

Lastly, to offer even more fine-grained control, the TNG hooks context that is returned with each call, will be an event emitter. It will fire events for each state change, as well as for each effect and cleanup.

Again, quoting an example from #20:

```js
function foo(val = 3) {
   var [x,updateX] = useState(val);
   updateX(v => v + 1);
   useEffect(function(){
      updateX(v => v + 1);
      console.log(`x is now: ${x + 2}`);
      return function(){ console.log(`cleaning things up: ${x + 2}`); };
   });
}

function onStateChange(hooksContext,hookIdx,prevValue,newValue) {
   console.log("** state:",hookIdx,prevValue,newValue);
}

function onEffect(hooksContext,effectIdx,effectFn) {
   console.log("** effect:",effectIdx);
}

function onCleanup(hooksContext,cleanupIdx,cleanupFn) {
   console.log("** cleanup:",cleanupIdx);
}

var context = foo().subscribe({ state: onStateChange, effect: onEffect, cleanup: onCleanup, });
// ** state: 0 3 4

// trigger pending effects...
context.effects();
// x is now: 5
// ** effect: 0
// ** state: 0 4 5

context = foo(context).subscribe(context.listeners);
// ** state: 0 5 6

// trigger pending effects...
context.effects();
// cleaning things up: 5
// ** cleanup: 0
// x is now: 7
// ** effect: 0
// ** state: 0 6 7

context.reset();
// cleaning things up: 7
// ** cleanup: 0
```

As you can see, right now there's the plan to have 3 classes of events: "state", "effect", and "cleanup". But obviously, as we get more types of effects, there will also be more types of events to subscribe to.

----

Bottom line: I think, based on what I believe you're asking, that TNG will have all the pieces you will want and need to fully control DOM/UI rendering and behavior.

But will TNG dictate a specific way that your DOM/UI must work, the way React does? **No.** That should be a separate project on top of TNG.

**Comment by WebReflection** [comment]
I'm asking for `useEffect(callback, [])` which invokes the callback only when the related node is live, and invoking the clean-up once the related node is disconnected.

[Live demo here](https://codepen.io/WebReflection/pen/maQXwq?editors=0010) where you can read in console `"connected"` and if you drop that button from the DOM you'll see `"disconnected"` too, all `useEffect` driven.

That's [dom-augmentor](https://github.com/WebReflection/dom-augmentor#dom-augmentor) in a nutshell.

**Comment by getify** [comment]
How is it determined that an element has been added to the DOM and removed from the DOM? I tried looking at dom-augmentor but it wasn't obvious to me from a brief read-through.

But assuming for a moment that there's a clean way to determine this and know about elements being added/removed... I would imagine a `useUIEffect(..)` or something like that, which sticks such an effect in a separate bucket than the basic `useEffect(..)` calls go.

Now, while the UI element is *connected*, you can call `.uiEffects()` (or whatever) to run those queued up effects. And if it's disconnected, you can call `.reset()` to run its pending cleanups.

Does that seem like it could work?

**Comment by getify** [comment]
I'm a bit confused still, but I'm trying to sort this out. I think it should be possible to accommodate what you're asking for, but I need more clarity.

I think what you're looking for is a way to "hook into" (or override) the part of `useEffect(..)` that decides whether to even schedule an effect -- normally just simply checking an array of input guards. And in this particular case, the way you want to decide or not is, if a DOM element is connected or not.

Do I have that summary correct so far?

If I understand correctly, from the perspective of `useEffect(..)`, if it receives a callback instead of an array of inputs, then it should invoke that callback. And I know that callback is listening to a DOM element's disconnected/connected events.

But where I get lost is, how does that callback then communicate back to `useEffect(..)` that yes, indeed, the effect in question should get scheduled?

And is that decision synchronous or asynchronous? By that I mean, can `useEffect(..)` know right away if the effect should be scheduled, or does it have to potentially wait until later to know the answer?

**Comment by WebReflection** [comment]
> I think what you're looking for is a way to "hook into" (or override) the part of `useEffect(..)` that decides whether to even schedule an effect

it doesn't have to be `useEffect` as it is, since I can wrap this library and provide my own `useEffect` for the DOM/UI, but yeah, I need a mechanism to "take over" the regular `useEffect` life cycle.

> how does that callback then communicate back to `useEffect(..)` that yes, indeed, the effect in question should get scheduled?

It doesn't. The callback takes over the behavior though the DOM. When the node connects, it invokes it, when it disconnect, it invokes its cleanup.

Bear in mind that an empty array is a very specific use case.


> And is that decision synchronous or asynchronous?

it's event driven, but always after any TNG stack 'cause `useEffect` is always **after** anyway.

> By that I mean, can `useEffect(..)` know right away if the effect should be scheduled

Nope, it has to do with the node/DOM content, as it's in component did mount.

> does it have to potentially wait until later to know the answer?

It doesn't have to wait anything, it has to delegate the in/out.

```js
[example] = TNG(example);

const btn = example();

setTimeout(
  // will add live attribute in Math.random() * 1000 seconds
  () => document.body.appendChild(btn),
  Math.random() * 1000
);

setTimeout(
  // will remove live attribute in 5 seconds
  () => document.body.removeChild(btn),
  5000
);

function example() {
  const {current:button} = useRef(Button);
  useEffect(() => {
    button.setAttribute('live');
    return () => {
      button.removeAttribute('live');
    };
  });
  return button;
}

function Button() {
  return document.createElement('button');
}
```

I hope the example is clear

**Comment by WebReflection** [comment]
To simplify, I've added a [Code Pen example](https://codepen.io/WebReflection/pen/YBENPK?editors=0010) that does the following:

```js
const {default: $, useRef, useEffect} = augmentor;

[example] = [$(example)];

const btn = example();

setTimeout(
  () => document.body.appendChild(btn),
  Math.random() * 1000
);

setTimeout(
  () => document.body.removeChild(btn),
  5000
);

function example() {
  const {current:button} = useRef(Button);
  useEffect(() => {
    button.textContent = 'online';
    console.log(button.textContent);
    return () => {
      button.textContent = 'offline';
      console.log(button.textContent);
    };
  }, []);
  return button;
}

function Button() {
  return document.createElement('button');
}
```

**Comment by getify** [comment]
What about something sorta like this:

```js
function hitCounter() {
   useEffect(function() { .. }, [ .. ]);
   useDOMEffect(function() { .. }, [ .. ]);
}

hitCounter = TNG(hitCounter);

var hooksContext = hitCounter();
hooksContext.effects();  // run all the standard effects now

Observer.addEventListener("connected", function() {
   hooksContext.DOMeffects();  // only run these DOM effects if the element is connected
} );
```

In other words, these "DOM effects" would always be "scheduled" onto the hooks-context object just like standard effects, but since they're still lazy they don't automatically run. You would conditionally decide to run them based on the "connected" state, etc.

Would something like that work at all?

**Comment by WebReflection** [comment]
exporting effects in a storable way might be useful, yes, but I need to play around with it to see how that actually works.

Running all the standard effects makes no sense though, s I wouldn't care/know about those.

```js
var hooksContext = hitCounter();
var domEffects = hooksContext.DOMeffects();

Observer.addEventListener("connected", domEffects.in);
Observer.addEventListener("disconnected", domEffects.out);
```

That would be definitively more user friendly.

### Still missing

Where the hack is the returned node/facade/dom to use in order to actually listen to anything?

Remember, this is about DOM/UI lifecycle, I need a node reference or I cannot observe a thing there.

**Comment by WebReflection** [comment]
TL;DR this is the most important contract of them all

```js
function example() {
  const {current:button} = useRef(Button);
  useEffect(() => {
    button.textContent = 'online';
    console.log(button.textContent);
    return () => {
      button.textContent = 'offline';
      console.log(button.textContent);
    };
  }, []);
  // return an observable "something" (DOM node) or GTFO
  return button;
}
```

**Comment by getify** [comment]
> exporting effects in a storable way might be useful, yes

A critical part of the future design of TNG (per #20) is that, to be more FP friendly, all the "state changes" and "side effects" that a invocation of the articulated function produces are encapsulated in the returned hooks-context object, instead of being internally encapsulated in that function itself. You get back a hooks-context object, and then you do whatever you want next (throw it away, or keep it as a starting point for the next invocation, run its various effects, or ignore them, etc etc etc). That way, the articulated function was, in essence, itself "pure".

That's why I was struggling to understand the idea of some other part of the system, like an external event, being able to schedule an effect on/inside the hooks context.

> Where the hack is the returned node/facade/dom to use in order to actually listen to anything?

The way it would work would be:

```js
function hitCounter() {
   var myBtn = ..;
   // ..
   return myBtn;
}

hitCounter = TNG(hitCounter);

var hooksContext = hitCounter();
hooksContext.value;   // myBtn
```

**Comment by WebReflection** [comment]
@getify this is not helping ... I came here to expand the limits of this project and so far, all I've seen, is some excuse for how this project it.

As summary, whenever you think this project, born after hooks that were born out of UI needs, will be capable of handling DOM events, please give me a shout.

**Comment by getify** [comment]
I'm sorry you felt that way. I wasn't arguing anything, just trying in good faith tounderstand your needs and try to figure how to weave them in or how to expand TNG to suit.

I am already using hooks for UI needs, and I expect to use them even more for UI things once the changes in #20 are implemented. I hope you'll reconsider or re-explore once those changes land. I certainly hope TNG can be useful for these sorts of things.

**Comment by WebReflection** [comment]
> just trying in good faith tounderstand your needs

that's your first mistake, I am asking nothing more than what my library can do already, and React too.


> I am already using hooks for UI needs

I don't think so, otherwise a working `useEffect` would be sufficient.


> the changes in #20 are implemented

I think nothing in #20 is relevant here

> I hope you'll reconsider or re-explore once those changes land. I certainly hope TNG can be useful for these sorts of things.

That's why I am here. I've solved already everything I'm asking for, but this library doesn't catch up.

As tweeted already, I wish there was one bus stop for this alternative, but the doesn't look like a possibility, if there's not much dialog, just some pre-defined statement of what this library is, and will ever been.

**Comment by getify** [comment]
I'm locking this conversation now, since your tone is insulting rather than trying to collaborate in good faith.
